src/syn/io/der.rs

```diff
@@ -145,62 +154,245 @@ impl Reader for DerReader {
+            vec.push(T::read_value(self)?);
+        }
+        Ok(vec)
+    }
+
+    fn read_set<C: set::Constraint, S: Sized, F: Fn(&mut Self) -> Result<S, Self::Error>>(&mut self, f: F) -> Result<S, Self::Error> {
+        self.read_sequence::<C, S, F>(f)
+    }
+
+    fn read_set_of<C: setof::Constraint, T: ReadableType>(&mut self) -> Result<Vec<<T as ReadableType>::Type>, Self::Error> {
+        self.read_sequence_of::<C, T>()
     }
 
     #[inline]
     fn read_enumerated<C: enumerated::Constraint>(&mut self) -> Result<C, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[enumerated] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        if let Length::Definite(l) = length {
+            let index = self.buffer.read_i64_number(l)? as u64;
+            C::from_choice_index(index).ok_or(Error::InvalidChoiceIndex(index, C::STD_VARIANT_COUNT))
+        } else {
+            Err(Error::UnsupportedOperation("Indefinite range is not supported in DER".to_string()))
+        }
     }
 
     #[inline]
     fn read_choice<C: choice::Constraint>(&mut self) -> Result<C, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        eprintln!("[choice] Class = {:#?}, PC = {:#?}, Tag = {:#?}, Length = {:#?}", class, pc, tag, length);
+
+        let index = tag as u64;
+
+        if let Class::ContextSpecific = &class {
+        } else {
+            return Err(Error::InvalidType)
+        }
+
+        if index >= C::STD_VARIANT_COUNT {
+            return Err(Error::InvalidChoiceIndex(index, C::VARIANT_COUNT))
+        }
+
+        Ok((index, C::read_content(index, self)?)).and_then(|(index, content)| {
+            content.ok_or_else(|| Error::InvalidChoiceIndex(index, C::VARIANT_COUNT))
+        })
     }
 
     #[inline]
     fn read_opt<T: ReadableType>(
         &mut self,
     ) -> Result<Option<<T as ReadableType>::Type>, Self::Error> {
-        unimplemented!()
+        let read_position = self.buffer.read_position;
+        match T::read_value(self) {
+            Ok(result) => Ok(Some(result)),
+            Err(Error::InvalidType) => {
+                self.buffer.read_position = read_position;
+                Ok(None)
+            },
+            Err(err) => Err(err)
+        }
     }
 
     #[inline]
     fn read_number<T: numbers::Number, C: numbers::Constraint<T>>(
         &mut self,
     ) -> Result<T, Self::Error> {
-        unimplemented!()
+        let (class, pc, tag) = self.buffer.read_identifier()?;
+        let length = self.buffer.read_length()?;
+
+        if let Class::Universal = &class {
+            if let (PC::Primitive, 2) = (&pc, &tag) {} else {
+                return Err(Error::InvalidType)
+            }
+        }
```

Comment:
With `asn1rs::model::Type` this can be collapsed to
```rust
if !matches!(tag, Tag::Universal(2)) || PC::Primitive != pc {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_INTEGER != tag || PC::Primitive != pc {
    return Err(Error::InvalidType);
}
```